lib/internal/wasi.js

```diff
@@ -0,0 +1,1301 @@
+        rightsBase: RIGHTS_SOCKET_BASE,
+        rightsInheriting: RIGHTS_SOCKET_INHERITING,
+      };
+    case stats.isFile():
+      return {
+        filetype: WASI_FILETYPE_REGULAR_FILE,
+        rightsBase: RIGHTS_REGULAR_FILE_BASE,
+        rightsInheriting: RIGHTS_REGULAR_FILE_INHERITING,
+      };
+    case stats.isSocket():
+      return {
+        filetype: WASI_FILETYPE_SOCKET_STREAM,
+        rightsBase: RIGHTS_SOCKET_BASE,
+        rightsInheriting: RIGHTS_SOCKET_INHERITING,
+      };
+    case stats.isSymbolicLink():
+      return {
+        filetype: WASI_FILETYPE_SYMBOLIC_LINK,
+        rightsBase: 0n,
+        rightsInheriting: 0n,
+      };
+    default:
+      return {
+        filetype: WASI_FILETYPE_UNKNOWN,
+        rightsBase: 0n,
+        rightsInheriting: 0n,
+      };
+  }
+};
+
+const stat = (wasi, fd) => {
+  const entry = wasi.FD_MAP.get(fd);
+  if (!entry) {
+    throw WASI_EBADF;
+  }
+  if (entry.filetype === undefined) {
+    const stats = fs.fstatSync(entry.real);
+    const {
+      filetype, rightsBase, rightsInheriting,
+    } = translateFileAttributes(fd, stats);
+    entry.filetype = filetype;
+    if (entry.rights === undefined) {
+      entry.rights = {
+        base: rightsBase,
+        inheriting: rightsInheriting,
+      };
+    }
+  }
+  return entry;
+};
+
+class WASI {
+  constructor({ preopenDirectories = { '.': '.' }, env = {}, args = [] } = {}) {
+    this.memory = undefined;
+    this.view = undefined;
+
+    this.FD_MAP = new Map([
+      [WASI_STDIN_FILENO, {
+        real: 0,
+        filetype: undefined,
+        rights: undefined,
+        path: undefined,
+      }],
+      [WASI_STDOUT_FILENO, {
+        real: 1,
+        filetype: undefined,
+        rights: undefined,
+        path: undefined,
+      }],
+      [WASI_STDERR_FILENO, {
+        real: 2,
+        filetype: undefined,
+        rights: undefined,
+        path: undefined,
+      }],
+    ]);
+
+    for (const [k, v] of Object.entries(preopenDirectories)) {
+      const real = fs.openSync(v);
+      const newfd = [...this.FD_MAP.keys()].reverse()[0] + 1;
+      this.FD_MAP.set(newfd, {
+        real,
+        filetype: WASI_FILETYPE_DIRECTORY,
+        rights: {
+          base: RIGHTS_DIRECTORY_BASE,
+          inheriting: RIGHTS_DIRECTORY_INHERITING,
+        },
+        fakePath: k,
+        path: v,
+      });
+    }
+
+    const getiovs = (iovs, iovsLen) => {
+      // iovs* -> [iov, iov, ...]
+      // __wasi_ciovec_t {
+      //   void* buf,
+      //   size_t buf_len,
+      // }
+
+      this.refreshMemory();
+
+      const buffers = Array.from({ length: iovsLen }, (_, i) => {
+        const ptr = iovs + (i * 8);
+        const buf = this.view.getUint32(ptr, true);
+        const bufLen = this.view.getUint32(ptr + 4, true);
+        return new Uint8Array(this.memory.buffer, buf, bufLen);
+      });
+
+      return buffers;
+    };
+
+    const CHECK_FD = (fd, rights) => {
+      const stats = stat(this, fd);
```

Comment:
Doing this for every `CHECK` call below seems pretty expensive… is there a way around that? Can we cache `rights`?

Reply:
`stat` itself caches